From 8db0c55c763a104ecf45083be23c99904c7de617 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Mon, 26 Jul 2021 18:20:55 +0800
Subject: [PATCH 21/23] qwaylandwindow: Clear mWaitingToApplyConfigure when
 reset

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 src/client/qwaylandwindow.cpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/client/qwaylandwindow.cpp b/src/client/qwaylandwindow.cpp
index 669bddf2..9ee05a45 100644
--- a/src/client/qwaylandwindow.cpp
+++ b/src/client/qwaylandwindow.cpp
@@ -232,6 +232,8 @@ bool QWaylandWindow::shouldCreateSubSurface() const
 
 void QWaylandWindow::reset()
 {
+    mWaitingToApplyConfigure = false;
+
     delete mShellSurface;
     mShellSurface = nullptr;
     delete mSubSurfaceWindow;
-- 
2.20.1

